File: SOLID/ISP/Compliant/isp_example.py
from abc import ABC, abstractmethod

class Payable(ABC):
    @abstractmethod
    def pay(self, amount): pass


class Refundable(ABC):
    @abstractmethod
    def refund(self, amount): pass


class RecurringPayable(ABC):
    @abstractmethod
    def setup_recurring(self, amount): pass


class UpiPayment(Payable, Refundable):
    def pay(self, amount):
        print(f"Paid ₹{amount} via UPI")

    def refund(self, amount):
        print(f"Refunded ₹{amount} via UPI")


class CreditCardPayment(Payable, Refundable, RecurringPayable):
    def pay(self, amount):
        print(f"Paid ₹{amount} via Credit Card")

    def refund(self, amount):
        print(f"Refunded ₹{amount} to Credit Card")

    def setup_recurring(self, amount):
        print(f"Recurring payment of ₹{amount} set up")

File: SOLID/ISP/Compliant/isp_example.ts
interface Payable {
  pay(amount: number): void;
}

interface Refundable {
  refund(amount: number): void;
}

interface RecurringPayable {
  setupRecurring(amount: number): void;
}

class UpiPayment1 implements Payable, Refundable {
  pay(amount: number) {
    console.log(`Paid ₹${amount} via UPI`);
  }

  refund(amount: number) {
    console.log(`Refunded ₹${amount} via UPI`);
  }
}

class CreditCardPayment implements Payable, Refundable, RecurringPayable {
  pay(amount: number) {
    console.log(`Paid ₹${amount} via Credit Card`);
  }

  refund(amount: number) {
    console.log(`Refunded ₹${amount} to Credit Card`);
  }

  setupRecurring(amount: number) {
    console.log(`Recurring payment of ₹${amount} set up`);
  }
}
